### PR TITLE
Add HBA config file support, add hba reload logic.

### PR DESCRIPTION
**Pull request by vadv**
add hba support, based on https://github.com/yandex/odyssey/pull/276.

### CODE REVIEW

**Comment by chobostar** [comment]
Dear maintainers @reshke @usernamedt @x4m !

I wonder if you could take a look to this PR? If you have a time, of course.

---

Despite, it might be looks "ugly", but the implementation almost the same in pgbouncer:
https://github.com/pgbouncer/pgbouncer/blob/pgbouncer_1_17_0/src/client.c#L237-L241

The purpose of this PR:
- the same feature as in pgbouncer, it could soften migration from pgbouncer to odyssey usage
- consistent and clear way to support access to pooler

p.s. IMO the performance test with/without that feature would be a nice.

File: sources/hba_reader.c [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+	od_keyword("hostssl", OD_LHOSTSSL),
+	od_keyword("hostnossl", OD_LHOSTNOSSL),
+	/* db/user */
+	od_keyword("all", OD_LALL),
+	od_keyword("sameuser", OD_LSAMEUSER),
+	/* auth type */
+	od_keyword("allow", OD_LALLOW),
+	od_keyword("trust", OD_LALLOW),
+	od_keyword("deny", OD_LDENY),
+	od_keyword("reject", OD_LDENY),
```
[the remark this answers is not in the document]
Comment (vadv) [reply sources/hba_reader.c:36]: Yes, that's right. For example, the keyword "trust", which is used in vanilla postgresql hba, means that the user can connects without authorization, so it seems to me that it is correct to stay compatibility with vanilla postgresql, but in fact it is correct to call it as the keyword "allow".

"deny" is done by analogy with hosts.allow and hosts.deny.

File: sources/hba.c [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+void od_hba_unlock(od_hba_t *hba)
+{
+	pthread_rwlock_unlock(&hba->lock);
+}
+
+void od_hba_reload(od_hba_t *hba, od_hba_rules_t *rules)
+{
+	od_hba_write_lock(hba);
```
[the remark this answers is not in the document]
Comment (vadv) [reply sources/hba.c:43]: We want to have a "read" lock to setup of new connection.

File: sources/hba_reader.c [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+	return token->type;
+}
+
+static int od_hba_reader_match_string(od_token_t token, char **value)
+{
+	char *copy = malloc(token.value.string.size + 1);
+	if (copy == NULL) {
+		return -1;
+	}
+	memcpy(copy, token.value.string.pointer, token.value.string.size);
```
[the remark this answers is not in the document]
Comment (vadv) [reply sources/hba_reader.c:130]: this is for config parser, why do we need to take into account null terminator?

[the remark this answers is not in the document]
Comment (reshke) [reply sources/hba_reader.c:130]: because you do it it line 131

File: sources/hba.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (vadv) [reply sources/hba.c]: changed to mutex

**Review by x4m — APPROVED** [review]

File: docker/entrypoint.sh [hunk 4]
```diff
@@ -20,17 +20,23 @@ then
 	exit 1
 fi
 
-ody-stop
```
Comment (x4m) [inline docker/entrypoint.sh:23]: Errr, does it work? I think we need to stop old instance, right?

Comment (x4m) [reply docker/entrypoint.sh:23]: Eithrer way we can remove stop at the end of the file :)

Comment (reshke) [reply docker/entrypoint.sh:23]: This change is ok, ody-stop is called in every test

**Comment by reshke** [comment]
LGTM

File: sources/hba_reader.c [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+		token->type = OD_PARSER_KEYWORD;
+		token->line = parser->line;
+		token->value.string.pointer = parser->pos;
+		while (parser->pos < parser->end && *parser->pos != ',' &&
+		       (isalnum(*parser->pos) || ispunct(*parser->pos)))
+			parser->pos++;
+		token->value.string.size =
+			parser->pos - token->value.string.pointer;
+		return token->type;
+	}
```
Comment (reshke) [inline sources/hba_reader.c:94]: so this is keyword parsing, why is it differs from od_parser_next keyword parsing?

```
        /* keyword */
        if (isalpha(*parser->pos)) {
                token->type = OD_PARSER_KEYWORD;
                token->line = parser->line;
                token->value.string.pointer = parser->pos;
                while (parser->pos < parser->end &&
                       (*parser->pos == '_' || isalpha(*parser->pos) ||
                        isdigit(*parser->pos)))
                        parser->pos++;
                token->value.string.size =
                        parser->pos - token->value.string.pointer;
                return token->type;
        }
```

File: sources/hba.c [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+bool od_hba_validate_addr6(od_hba_rule_t *rule, struct sockaddr_storage *sa)
+{
+	struct sockaddr_in6 *sin = (struct sockaddr_in6 *)sa;
+	struct sockaddr_in6 *rule_addr = (struct sockaddr_in6 *)&rule->addr;
+	struct sockaddr_in6 *rule_mask = (struct sockaddr_in6 *)&rule->mask;
+	for (int i = 0; i < 16; ++i) {
+		uint8_t client_net_byte = rule_mask->sin6_addr.s6_addr[i] &
+					  sin->sin6_addr.s6_addr[i];
+		if (client_net_byte ^ rule_addr->sin6_addr.s6_addr[i]) {
```
Comment (reshke) [inline sources/hba.c:64]: Such a strange way to write x == y ?

**Review by reshke — APPROVED** [review]

**Comment by reshke** [comment]
Thanks to @sanyabas for initial implementation and @vadv for putting effort on this to be merged, also thanks @chobostar for review

**Pull request merged**